treewide: Use hyphens instead of underscores for property separation

rruuaanng opened the pull request:
This inherits from proposal #53502, with the main work as shown in the title. For details, see https://github.com/zephyrproject-rtos/zephyr/issues/53506.


Closes: #83741

rruuaanng commented:
I have a lot of work to catch up on right now, like CI checks and other underlying changes (maybe due to a lack of focus).

rruuaanng commented:
I wrote another PR which adds the devicetree coding style to the contributor guide.
(it provides a link to the upstream Linux kernel devicetree coding style.)

rruuaanng commented:
I have marked this PR as ready for review, and I sincerely apologize for this action. Even though it was well-received by most in 2023, I now recognize the need to heed the advice of the new maintainers. for style checks. (the primary concern, and indeed the final issue), is whether we should delegate this to CI for enforcement or rely on developers to adhere to it, with maintainers ensuring its smooth implementation.

kartben reviewed (dismissed):
This is a big breaking change that needs to be called out in migration guide. It can be done using a generic sentence rather than explicitly listing the changes for all the bindings impacted.

It *might* work to deprecate the old properties, which would be better IMO. dtlib/edtlib might complain about it, I guess, but worth trying (might even be worth implementing a fix to allow potentially clashing deprecated/non deprecated properties to co-exist)

rruuaanng commented:
> This is a big breaking change that needs to be called out in migration guide. It can be done using a generic sentence rather than explicitly listing the changes for all the bindings impacted.

I will add a note later regarding the migration.

> dtlib/edtlib might complain about it, I guess, but worth trying (might even be worth implementing a fix to allow potentially clashing deprecated/non deprecated properties to co-exist)

We don’t need to modify them. The parser only attempts to find matching items when handling binding dependencies, so we just need to keep the changes consistent. (This PR behaves similarly, and CI looks good.)

kartben commented:
> We don’t need to modify them. The parser only attempts to find matching items when handling binding dependencies, so we just need to keep the changes consistent. (This PR behaves similarly, and CI looks good.)

Not sure what you mean. Downstream users of these bindings will see their code fail for now they will potentially miss required properties as they would still be using their old _ names instead of the new ones

cfriedt commented:
To compliment this Ideally, we would have something like

* a `.github/workflows/dt-underscores.yml` that would prevent new bindings from being added with underscores
* a new python file in `scripts/` that could be used by downstream users to automatically correct DT files of all kinds.
* a Kconfig option, default y, which will be deprecated next release, that will automatically convert underscores to hyphens e.g. with an edtlib option. This should probably print a warning when it corrects underscores

rruuaanng commented:
Ok, it seems like the direction is right, I will try to make this change satisfy all the changes.

rruuaanng commented:
Remaining work:
1. a .github/workflows/dt-underscores.yml that would prevent new bindings from being added with underscores
2. a Kconfig option, default y, which will be deprecated next release, that will automatically convert underscores to hyphens e.g. with an edtlib option. This should probably print a warning when it corrects underscores


Pending:
> a new python file in scripts/ that could be used by downstream users to automatically correct DT files of all kinds.

According to my investigation, the script would require modifications to all bindings and device trees, replacing underscores, which is not easy to implement. For this request, perhaps we should delay it. As mentioned below:

> @galak mentions that out of tree users would benefit from a script that helps with the transition from _ to -, otherwise he's in favor. The script might not be trivial and is unclear whether it is worthwhile.

rruuaanng commented:
@cfriedt  I think we can provide a small feature in `check_compliance.py` that, after running, prints the binding paths of properties containing underscores to the console, allowing users to modify them on their own without having to go through the trouble of searching for them.

rruuaanng commented:
Okay, everyone, we can revisit it. If there are no other issues, it can be implemented soon, allowing users to make adjustments right away.

rruuaanng commented:
Change log:

When building, if the devicetree contains an underlined attribute, the build will stop and output the binding path and devicetree source file path that need to be changed.

decsny reviewed (dismissed):
the git history is not bisectable if the bindings are changed in one commit and dts changed to match in a later commit

rruuaanng commented:
> the git history is not bisectable if the bindings are changed in one commit and dts changed to match in a later commit

Great! I'll consolidate them into a single commit later.

cfriedt commented:
@rruuaanng - it might make sense to keep this PR in draft until it's passing all tests and has sufficient early feedback.

rruuaanng commented:
Maybe keeping it open will attract more people to discuss it. Until then, I won't make any more changes. Thank you for your continued follow-up :)

rruuaanng commented:
Change log:

Add style transfer script `migrate_bindings_style.py`.

rruuaanng commented:
@cfriedt We can now confidently proceed with this proposal, enabling users to easily migrate through scripts and providing support for user-defined directories.

The final step is to add an @mbolivar sign-off to the commit, specifically for the `edtlib` and `CI checker` sections. For these, I have referred to his implementation, and this needs to be accompanied by his personal authorization.

cfriedt commented:
@rruuaanng - you should try to fix compliance issues and add licenses where they are missing.

To mirror the checks in CI locally, use

```shell
./scripts/checkpatch.pl -g main..
```
and

```shell
./scripts/ci/check_compliance.py -c main..
```

https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style

cfriedt commented:
@rruuaanng - it would be great to separate the very large change of all dts files to other PRs. Currently it's very difficult to review this PR.

Can I suggest maybe including only one or two dts changes here to verify that the scripts and GitHub actions are functioning properly? The rest of the dts changes should be made in separate PRs, possibly grouped by areas listed in the maintainers file.

See
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html

rruuaanng commented:
I agree, I will separate the huge DTS commit into another PR later, it makes my force push very slow.

rruuaanng commented:
@cfriedt The advancement of this proposal requires that all sub-PRs be merged and this PR be processed last. Because CI will throw a warning, we need to leave it to the end to solve it. They appear to have dozens of commits that modify the bindings and DTS between different subsystems.

rruuaanng commented:
I believe I need to open this issue to the community, as it is a large-scale task. Given my current workload, I’m unable to dedicate my full attention to it. I will focus solely on handling this PR.

cfriedt commented:
> @cfriedt The advancement of this proposal requires that all sub-PRs be merged and this PR be processed last. Because CI will throw a warning, we need to leave it to the end to solve it. They appear to have dozens of commits that modify the bindings and DTS between different subsystems.

Call me skeptical (hopeful?), but I think that there is probably a way to a) activate CI checks only when a DT binding has changed, and b) find one binding to change that does not require changing the entire tree all at once.

rruuaanng commented:
> > @cfriedt The advancement of this proposal requires that all sub-PRs be merged and this PR be processed last. Because CI will throw a warning, we need to leave it to the end to solve it. They appear to have dozens of commits that modify the bindings and DTS between different subsystems.
> 
> Call me skeptical (hopeful?), but I think that there is probably a way to a) activate CI checks only when a DT binding has changed, and b) find one binding to change that does not require changing the entire tree all at once.

Oh, I almost forgot. Currently, the binding style checks are skipped when no changes are made to the bindings. Let's wait and see.

rruuaanng commented:
> > The final step is to add an @mbolivar sign-off to the commit, specifically for the `edtlib` and `CI checker` sections. For these, I have referred to his implementation, and this needs to be accompanied by his personal authorization.
> 
> @rruuaanng thanks again for taking this on. Just to clarify, you don't need my personal authorization to add my signed-off-by: to any commits that you create, **as long as I already added the signed-off-by: myself and opened a pull request with that commit to upstream zephyr**.
> 
> When I added my signed-off-by: and created a PR, I was certifying for legal purposes (see the [DCO](https://docs.zephyrproject.org/latest/contribute/guidelines.html#developer-certification-of-origin-dco) docs) that I was able to submit the code under the appropriate license (apache 2 or other file-specific license) to the project at the time I opened the PR.
> 
> You are free to use the code under that license however you want, as long as you comply with the license's terms. You don't need any additional permission from me. The only thing you have to do is make sure that my signed-off-by: lines stay on any commits you submit yourself if they are adapted from my work.

I hope I haven't misunderstood. This is my current commit message:
```
    dts: bindings: Change the property names in the bindings

    Change the property names in the bindings and DTS to use
    hyphens(-) for separation instead of underscores(_).

    Signed-off-by: James Roy <rruuaanng@outlook.com>
    Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
```

This commit is based on your work, so I should change it to the following:
```
    dts: bindings: Change the property names in the bindings

    Change the property names in the bindings and DTS to use
    hyphens(-) for separation instead of underscores(_).

    Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
```

But for content I have written myself, I only need to use my own signature:
```
    scripts: Add binding migration script

    This script replaces all underscore(_)
    separation in a binding with hyphens(-).

    Signed-off-by: James Roy <rruuaanng@outlook.com>
```

rruuaanng commented:
CI throw
```
 561f94309a80cfa060b1f38e896f273adb4ade2c: author email (James Roy <rruuaanng@outlook.com>) needs to match one of the signed-off-by entries.
Compliance error, check for error messages in the "Run Compliance Tests" step
```
Okay, it looks like only the signature issue remains. It seems like there might be a misunderstanding on my part? I might need to keep the original commit message.

from https://docs.zephyrproject.org/latest/contribute/guidelines.html#developer-certification-of-origin-dco
```
If you are altering an existing commit created by someone else, you must add your Signed-off-by: line without removing the existing one.
```
Hmmm.... It seems that the development process description isn't very clear on this. I apologize for not understanding it even after reading it multiple times. Does "keep" mean adding two signatures?

cfriedt commented:
It's definitely saying something about the bottom commit

<img width="937" alt="Screenshot 2025-01-09 at 10 23 01 PM" src="https://github.com/user-attachments/assets/a9c32c9e-d210-4cc2-bfd0-94ff93a75a57" />

cfriedt commented:
@rruuaanng - you should be able to do `git rebase -i HEAD~3`, enter `r` instead of `pick` for that commit, and then update the commit message to add your signed-off-by line.

rruuaanng commented:
> @rruuaanng - you should be able to do `git rebase -i HEAD~3`, enter `r` instead of `pick` for that commit, and then update the commit message to add your signed-off-by line.

Should I add two signatures? Like this
```
    Signed-off-by: James Roy <rruuaanng@outlook.com>
    Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
```

kartben commented:
> > @rruuaanng - you should be able to do `git rebase -i HEAD~3`, enter `r` instead of `pick` for that commit, and then update the commit message to add your signed-off-by line.
> 
> Should I add two signatures? Like this
> ```
>     Signed-off-by: James Roy <rruuaanng@outlook.com>
>     Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
> ```

Yes

rruuaanng commented:
Sure, everything looks good with the CI, and we can make a decision on the proposal.

cfriedt commented on .ruff-excludes.toml:
```diff
@@ -435,6 +435,9 @@
     "I001",     # https://docs.astral.sh/ruff/rules/unsorted-imports
     "UP035",    # https://docs.astral.sh/ruff/rules/deprecated-import
 ]
+"./scripts/migrate_bindings_style.py" = [
+    "I001",     # https://docs.astral.sh/ruff/rules/unsorted-imports
+]
```
Probably can work around this using `ruff` - @pdgendt is fairly familiar with that.

cfriedt replied:
I think it's just `ruff format <filename>`

rruuaanng replied:
Done. I have removed it.

cfriedt commented on scripts/ci/check_compliance.py:
```diff
@@ -36,6 +37,28 @@
 import list_boards
 import list_hardware
 
+sys.path.insert(0, os.path.join(str(Path(__file__).resolve().parents[2]),
```
I'm surprised that so much needs to be added in this file

cfriedt commented on scripts/migrate_bindings_style.py:
```diff
@@ -0,0 +1,91 @@
+
+
+# bindings base
+BINDINGS_PATH = ["dts/bindings/"]
```
Probably want this to be a list of Path objects

rruuaanng replied:
Strings list are convenient for command lines to pass parameters, see line  84

rruuaanng commented:
They are scattered into a task list, and you can give some help when you have time :)
#83741

rruuaanng commented:
@kartben Since this PR already includes a migration entry indicating that Zephyr device trees and binding properties now use hyphenated names, do the remaining PRs in worklist #83741 still need to add their own migration entries?

rruuaanng commented:
Change log:

Reorganize the description of the migration document.

kartben commented:
@rruuaanng will you come back to this?

rruuaanng commented:
@kartben Yes, I'm ready to make all changes at any time.

cfriedt commented on scripts/ci/check_compliance.py:
```diff
@@ -36,6 +37,28 @@
 import list_boards
 import list_hardware
 
+sys.path.insert(0, os.path.join(str(Path(__file__).resolve().parents[2]),
+                                'scripts/dts/python-devicetree/src'))
```
I believe `os.path.join()` is not needed when using `Path` objects

```suggestion
sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "scripts" / "dts" / "python-devicetree" / "src")
```

rruuaanng replied:
Done.

cfriedt replied:
Has this been fixed?

rruuaanng replied:
Hmm... wait, why didn't my changes appear? I changed it again.

kartben commented on scripts/ci/bindings_propertys_name_wl.yaml:
propertys -> properties
Also need to use a more inclusive term than "whitelist". Recommendation is "allowlist"

rruuaanng replied:
I have changed all whitelist-like names to allowlist. +1

cfriedt commented on scripts/ci/check_compliance.py:
```diff
@@ -36,6 +37,28 @@
+    os.environ['ZEPHYR_BASE'] = ZEPHYR_BASE
+
+# Initialize the propertys name whitelist
+BINDINGS_PROPERTYS_WL = None
+with open(Path(__file__).parent / 'bindings_properties_allowlist.yaml', 'r') as f:
+    wl = yaml.safe_load(f.read())
+    if wl is not None:
+        BINDINGS_PROPERTYS_WL = set(wl)
+    else:
+        BINDINGS_PROPERTYS_WL = {'',}
```
```suggestion
BINDINGS_PROPERTYS_AL = None
with open(Path(__file__).parent / 'bindings_properties_allowlist.yaml', 'r') as f:
    al = yaml.safe_load(f.read())
    if al is not None:
        BINDINGS_PROPERTYS_AL = set(al)
    else:
        BINDINGS_PROPERTYS_AL = {'',}
```

cfriedt commented on scripts/migrate_bindings_style.py:
```diff
@@ -0,0 +1,90 @@
+
+# bindings base
+BINDINGS_PATH = [Path("dts/bindings/")]
+BINDINGS_PROPERTIES_ALLOWLIST = {
+    "mmc-hs400-1_8v",
+    "mmc-hs200-1_8v",
+}
```
Does this need to be initialized if we refer to the allow list in yaml?

rruuaanng replied:
I plan to modify it to reference the allowlist file in the ci directory (the reason I initially avoided this is because using files from the ci directory seemed somewhat inappropriate, as implied by the directory name). Perhaps I should move the allowlist file to the scripts directory instead, so it doesn’t appear out of place.

cfriedt replied:
Yeah, it would probably be ok to move it to the scripts directory.

cfriedt commented on scripts/ci/check_compliance.py:
```diff
@@ -348,32 +371,75 @@ class DevicetreeBindingsCheck(ComplianceTest):
+        Checks if the property names in the binding file contain underscores.
+        """
+        for prop_name in binding.prop2specs:
+            if '_' in prop_name and prop_name not in BINDINGS_PROPERTYS_WL:
```
```suggestion
            if '_' in prop_name and prop_name not in BINDINGS_PROPERTYS_AL:
```

carlescufi commented:
Hi @rruuaanng. Would you be willing and able to present this in the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group)? If so, please let me know of your availability here and I will schedule a date for you.

rruuaanng commented:
@carlescufi My spoken English isn't that great, and I might miss some parts of the conversation. I think I could organize a PPT to showcase my work, and I'll upload the file to my personal repository. If permitted, I'll start preparing it :) WDYT?

cfriedt reviewed (dismissed):
Looks ok to me

mbolivar reviewed (commented):
@rruuaanng this PR looks incomplete to me -- sorry if I missed some discussion in the thread, but the title still says it's about the treewide conversion. The current PR contains some conversion and CI checks, but not the actual changes themselves. I'm not blocking this, but I'm confused about its status

rruuaanng commented:
> but not the actual changes themselves

If you're referring to the changes related to dts and bindings, I have moved them to #83741. Since there were too many, I needed to separate them to make the review process smoother.

cfriedt:
> Can I suggest maybe including only one or two dts changes here to verify that the scripts and GitHub actions are functioning properly? The rest of the dts changes should be made in separate PRs, possibly grouped by areas listed in the maintainers file.

rruuaanng commented:
Oh, I accidentally performed some magical operations while using the merge conflict tool. I apologize for the noise it caused.

cfriedt commented:
Yeah, there is an error about an empty commit message.

rruuaanng commented:
Hmm, this is a bit confusing to me, it seems like the current PR description is the same as it was originally.

cfriedt commented:
IIRC the compliance issues are false positives.

rruuaanng committed:
scripts: Add bindings style migration script

This script replaces all underscore(_)
separation in a binding with hyphens(-).

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng committed:
scripts: ci: Add CI bindings style checker

Implement a check in the CI pipeline to enforce
that property names in device tree bindings do
not contain underscores.

Signed-off-by: James Roy <rruuaanng@outlook.com>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

rruuaanng committed:
doc: Add a new migration guide entry

Bindings now use hyphens(-) instead of
underscores(_) as property separators.

Signed-off-by: James Roy <rruuaanng@outlook.com>

maass-hamburg commented:
@rruuaanng some props were not migrated and are now causing ci fails, when dts yamls are changed, fix: 91818
